pyadjoint/optimization/tao_solver.py

```diff
@@ -0,0 +1,491 @@
+                self._shift += alpha
+
+            def mult(self, A, x, y):
+                M_dot = taoobjective.new_M()
+                from_petsc(x, M_dot)
+                ddJ = taoobjective.hessian(self._M, M_dot)
+                to_petsc(y, ddJ)
+                if self._shift != 0.0:
+                    y.axpy(self._shift, x)
+
+        H_matrix = PETSc.Mat().createPython(((n, N), (n, N)),
+                                            Hessian(), comm=comm)
+        H_matrix.setOption(PETSc.Mat.Option.SYMMETRIC, True)
+        H_matrix.setUp()
+        tao.setHessian(hessian, H_matrix)
+
+        class GradientNorm:
+            def mult(self, A, x, y):
+                X = taoobjective.new_M_dual()
+                from_petsc(x, X)
+                assert len(taoobjective.reduced_functional.controls) == len(X)
+                X = tuple(m._ad_convert_type(x, options=convert_options)
+                          for m, x in zip(taoobjective.reduced_functional.controls, X))
+                to_petsc(y, X)
+
+        M_inv_matrix = PETSc.Mat().createPython(((n, N), (n, N)),
+                                                GradientNorm(), comm=comm)
+        M_inv_matrix.setOption(PETSc.Mat.Option.SYMMETRIC, True)
+        M_inv_matrix.setUp()
+        tao.setGradientNorm(M_inv_matrix)
+
+        if problem.bounds is not None:
+            lbs = []
+            ubs = []
+            assert len(problem.bounds) == len(problem.reduced_functional.controls)
+            for (lb, ub), control in zip(problem.bounds, taoobjective.reduced_functional.controls):
+                if lb is None:
+                    lb = np.finfo(PETSc.ScalarType).min
+                if ub is None:
+                    ub = np.finfo(PETSc.ScalarType).max
+                lbs.append(lb)
+                ubs.append(ub)
+
+            lb_vec = vec_interface.new_petsc()
+            ub_vec = vec_interface.new_petsc()
+            to_petsc(lb_vec, lbs)
+            to_petsc(ub_vec, ubs)
+            tao.setVariableBounds(lb_vec, ub_vec)
+            lb_vec.destroy()
+            ub_vec.destroy()
+
+        options = PETScOptions(f"_pyadjoint__{tao.name}_")
+        options.update(parameters)
+        tao.setOptionsPrefix(options.options_prefix)
+
+        tao.setFromOptions()
+
+        if tao.getType() in {PETSc.TAO.Type.LMVM, PETSc.TAO.Type.BLMVM}:
+            class InitialHessian:
+                pass
+
+            class InitialHessianPreconditioner:
+                def apply(self, pc, x, y):
+                    X = taoobjective.new_M_dual()
+                    from_petsc(x, X)
+                    assert len(taoobjective.reduced_functional.controls) == len(X)
+                    X = tuple(m._ad_convert_type(x, options=convert_options)
+                              for m, x in zip(taoobjective.reduced_functional.controls, X))
+                    to_petsc(y, X)
+
+            B_0_matrix = PETSc.Mat().createPython(((n, N), (n, N)),
```

Comment:
Have mercy on your reader and explain that B_0 is H_0

Reply:
Added reference to Nocedal and Weight for the notation.